### ISSUE TITLE
Audioset labels(.json file)

### ISSUE THREAD

**Issue by sunyulin0421** [LudovicTuncay/hierarchical-label-propagation#1]
hi, @LudovicTuncay 

First of all, congratulations on your excellent scientific research results!

It is well known that Audioset datasets are formatted as **balanced training sets, unbalanced training sets, and validation sets**. Specifically, this typically includes audio files and json files for corresponding labels **(audioset_tr_full.json, audioset_eval_full.json)**.

However, the input and output formats of HLP are in the form of tensors, do you provide a script to convert the json file into the form of sparse tensors according to a mapping table, so as to obtain a tag file that conforms to the HLP input format?

Hope you can provide the method or script to use, thank you very much!

Sincerely wish you smooth sailing on the road of scientific research!

**Comment by LudovicTuncay** [comment]
We use **multi-hot encoding** to represent that an item can belong to multiple classes simultaneously by setting the corresponding class indices in a binary vector to 1, while all other indices remain 0.


In the case of AudioSet, the index of each label is specified in the [class_labels_indices.csv](http://storage.googleapis.com/us_audioset/youtube_corpus/v1/csv/class_labels_indices.csv) given in the [download page of the AudioSet dataset](https://research.google.com/audioset/download.html).  

### 1. From explicit labels to multi-hot

In the CSV (not sure if there are JSONs) files given in the download page, each example is associated with a list of positive labels (mids).

we can convert it to a list of positive labels' indices using:

```python
class_labels_indices_df = pd.read_csv("class_labels_indices.csv")
labels_str = "/m/068hy,/m/07q6cd_,/m/0bt9lr,/m/0jbk" # example taken from the first row of `eval_segments.csv`
labels = labels_str.split(",")
positive_labels_ids = []
for label in labels:
    index_value = int(class_labels_indices_df[class_labels_indices_df['mid'] == label].index[0])
    positive_labels_ids.append(index_value)

positive_labels_ids
>>> [73, 361, 74, 72]
```

and then compute the multi-hot encoding:

```python
classes_num = 527
multi_hot = torch.zeros(classes_num)
if positive_labels_ids: # check if the list is non-empty
    multi_hot[positive_labels_ids] = 1.0
```

You now can use **HLP** on that tensor:

```python
from hierarchical_label_propagation import HLP
hlp = HLP()
multi_hot_hlp = hlp.propagate(multi_hot)
```

### 2. From multi-hot to explicit label

You could also go in the other direction and extract the indices from the multi hot encoded tensor this way:
```python
def get_positive_indices(multi_hot_encoded):
    if multi_hot_encoded.ndim == 1:
        return torch.where(multi_hot_encoded == 1)[0].tolist()
    elif multi_hot_encoded.ndim == 2:
        return [torch.where(row == 1)[0].tolist() for row in multi_hot_encoded]
    else:
        raise ValueError("Input tensor must be 1D or 2D.")

positive_indices = get_positive_indices_torch(your_multihot_tensor)
```

For instance, using the multi-hot tensor from before, you would have:
```python
positive_indices = get_positive_indices(multi_hot)
positive_indices
>>> [72, 73, 74, 361] # which is the same as before, but sorted 

positive_indices_hlp = get_positive_indices(multi_hot_hlp)
positive_indices_hlp
>>> [72, 73, 74, 354, 361] # Using HLP added the label #361
```

You can then further process these lists to get the original label representation (mid and/or display names):

```python
def get_labels_from_indices(indices_output, class_labels_df):
    if isinstance(indices_output[0], list):  # Case for 2D input
        all_mids = []
        all_display_names = []
        for index_list in indices_output:
            mids = class_labels_df.loc[index_list, 'mid'].tolist()
            display_names = class_labels_df.loc[index_list, 'display_name'].tolist()
            all_mids.append(mids)
            all_display_names.append(display_names)
        return all_mids, all_display_names
    else:  # Case for 1D input
        mids = class_labels_df.loc[indices_output, 'mid'].tolist()
        display_names = class_labels_df.loc[indices_output, 'display_name'].tolist()
        return mids, display_names
```

Which, using the same examples as before would give:

```python
mids, display_names = get_labels_from_indices(positive_indices, class_labels_indices_df)
mids_hlp, display_names_hlp = get_labels_from_indices(positive_indices_hlp, class_labels_indices_df)

mids
>>> ['/m/0jbk', '/m/068hy', '/m/0bt9lr', '/m/07q6cd_'] # before HLP
mids_hlp
>>> ['/m/0jbk', '/m/068hy', '/m/0bt9lr', '/m/02dgv', '/m/07q6cd_'] # After HLP


display_names
>>> ['Animal', 'Domestic animals, pets', 'Dog', 'Squeak']
display_names_hlp
>>> ['Animal', 'Domestic animals, pets', 'Dog', 'Door', 'Squeak']
```

In this specific example, you can see that the label `Door` was added. This is because in the AudioSet dataset's subset of the ontology, Squeak only has 1 parent (contrary to the original ontology where it has 2 parents) which is door, therefore it is added. Door is a root node in that subset so the propagation ends here.

**Comment by sunyulin0421** [comment]
@LudovicTuncay Thanks,

> We use **multi-hot encoding** to represent that an item can belong to multiple classes simultaneously by setting the corresponding class indices in a binary vector to 1, while all other indices remain 0.
> 
> In the case of AudioSet, the index of each label is specified in the [class_labels_indices.csv](http://storage.googleapis.com/us_audioset/youtube_corpus/v1/csv/class_labels_indices.csv) given in the [download page of the AudioSet dataset](https://research.google.com/audioset/download.html).
> 
> ### 1. From explicit labels to multi-hot
> In the CSV (not sure if there are JSONs) files given in the download page, each example is associated with a list of positive labels (mids).
> 
> we can convert it to a list of positive labels' indices using:
> 
> class_labels_indices_df = pd.read_csv("class_labels_indices.csv")
> labels_str = "/m/068hy,/m/07q6cd_,/m/0bt9lr,/m/0jbk" # example taken from the first row of `eval_segments.csv`
> labels = labels_str.split(",")
> positive_labels_ids = []
> for label in labels:
>     index_value = int(class_labels_indices_df[class_labels_indices_df['mid'] == label].index[0])
>     positive_labels_ids.append(index_value)
> 
> positive_labels_ids
> >>> [73, 361, 74, 72]
> and then compute the multi-hot encoding:
> 
> classes_num = 527
> multi_hot = torch.zeros(classes_num)
> if positive_labels_ids: # check if the list is non-empty
>     multi_hot[positive_labels_ids] = 1.0
> You now can use **HLP** on that tensor:
> 
> from hierarchical_label_propagation import HLP
> hlp = HLP()
> multi_hot_hlp = hlp.propagate(multi_hot)
> ### 2. From multi-hot to explicit label
> You could also go in the other direction and extract the indices from the multi hot encoded tensor this way:
> 
> def get_positive_indices(multi_hot_encoded):
>     if multi_hot_encoded.ndim == 1:
>         return torch.where(multi_hot_encoded == 1)[0].tolist()
>     elif multi_hot_encoded.ndim == 2:
>         return [torch.where(row == 1)[0].tolist() for row in multi_hot_encoded]
>     else:
>         raise ValueError("Input tensor must be 1D or 2D.")
> 
> positive_indices = get_positive_indices_torch(your_multihot_tensor)
> For instance, using the multi-hot tensor from before, you would have:
> 
> positive_indices = get_positive_indices(multi_hot)
> positive_indices
> >>> [72, 73, 74, 361] # which is the same as before, but sorted 
> 
> positive_indices_hlp = get_positive_indices(multi_hot_hlp)
> positive_indices_hlp
> >>> [72, 73, 74, 354, 361] # Using HLP added the label #361
> You can then further process these lists to get the original label representation (mid and/or display names):
> 
> def get_labels_from_indices(indices_output, class_labels_df):
>     if isinstance(indices_output[0], list):  # Case for 2D input
>         all_mids = []
>         all_display_names = []
>         for index_list in indices_output:
>             mids = class_labels_df.loc[index_list, 'mid'].tolist()
>             display_names = class_labels_df.loc[index_list, 'display_name'].tolist()
>             all_mids.append(mids)
>             all_display_names.append(display_names)
>         return all_mids, all_display_names
>     else:  # Case for 1D input
>         mids = class_labels_df.loc[indices_output, 'mid'].tolist()
>         display_names = class_labels_df.loc[indices_output, 'display_name'].tolist()
>         return mids, display_names
> Which, using the same examples as before would give:
> 
> mids, display_names = get_labels_from_indices(positive_indices, class_labels_indices_df)
> mids_hlp, display_names_hlp = get_labels_from_indices(positive_indices_hlp, class_labels_indices_df)
> 
> mids
> >>> ['/m/0jbk', '/m/068hy', '/m/0bt9lr', '/m/07q6cd_'] # before HLP
> mids_hlp
> >>> ['/m/0jbk', '/m/068hy', '/m/0bt9lr', '/m/02dgv', '/m/07q6cd_'] # After HLP
> 
> 
> display_names
> >>> ['Animal', 'Domestic animals, pets', 'Dog', 'Squeak']
> display_names_hlp
> >>> ['Animal', 'Domestic animals, pets', 'Dog', 'Door', 'Squeak']
> In this specific example, you can see that the label `Door` was added. This is because in the AudioSet dataset's subset of the ontology, Squeak only has 1 parent (contrary to the original ontology where it has 2 parents) which is door, therefore it is added. Door is a root node in that subset so the propagation ends here.

@LudovicTuncay Thanks.
Wish you greater achievements on the road of scientific research.
